jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -191,42 +189,54 @@ public int getMaxCapacity()
     public RetainableByteBuffer acquire(int size, boolean direct)
     {
         RetainedBucket bucket = bucketFor(size, direct);
+
+        // No bucket, return non-pooled.
         if (bucket == null)
-            return newRetainableByteBuffer(size, direct, this::removed);
+            return newRetainableByteBuffer(size, direct, null);
+
+        // Try to acquire a pooled entry.
         Pool.Entry<RetainableByteBuffer> entry = bucket.getPool().acquire();
+        if (entry != null)
+        {
+            RetainableByteBuffer buffer = entry.getPooled();
+            ((Buffer)buffer).acquire();
+            return buffer;
+        }
+
+        // No pooled entry available, try to reserve.
+        Pool.Entry<RetainableByteBuffer> reservedEntry = bucket.getPool().reserve();
+        // Cannot reserve, return non-pooled.
+        if (reservedEntry == null)
+            return newRetainableByteBuffer(size, direct, null);
```

Comment:
move the reservation to release as well.  The behaviour is then similar (initially) our previous pools where a buffer was only pooled when returned

Reply:
On it.